service/storage_service.cc

```diff
@@ -1003,6 +1005,70 @@ future<> storage_service::raft_bootstrap(raft::server& raft_server) {
     }
 }
 
+future<> storage_service::update_topology_with_local_metadata(raft::server& raft_server) {
+    // TODO: include more metadata here
+    auto local_shard_count = smp::count;
+    auto local_ignore_msb = _db.local().get_config().murmur3_partitioner_ignore_msb_bits();
+
+    auto synchronized = [&] () {
+        auto it = _topology_state_machine._topology.find(raft_server.id());
+        if (!it) {
+            throw std::runtime_error{"Removed from topology while performing metadata update"};
+        }
+
+        auto& replica_state = it->second;
+
+        return replica_state.shard_count == local_shard_count
+            && replica_state.ignore_msb == local_ignore_msb;
+    };
+
+    // We avoid performing a read barrier if we're sure that our metadata stored in topology
+    // is the same as local metadata. Note that only we can update our metadata, other nodes cannot.
+    //
+    // We use a persisted flag `must_update_topology` to avoid the following scenario:
+    // 1. the node restarts and its metadata changes
+    // 2. the node commits the new metadata to topology, but before the update is applied
+    //    to the local state machine, the node crashes
+    // 3. then the metadata changes back to old values and node restarts again
+    // 4. the local state machine tells us that we're in sync, which is wrong
+    // If the persisted flag is true, it tells us that we attempted a metadata change earlier,
+    // forcing us to perform a read barrier even when the local state machine tells us we're in sync.
+
+    if (synchronized() && !(co_await _sys_ks.local().get_must_synchronize_topology())) {
+        co_return;
+    }
+
+    while (true) {
+        slogger.info("raft topology: performing read barrier");
```

Comment:
shouldn't this implementation detail be demoted to debug level logging?

Reply:
I made it INFO level because it's easier to see why Scylla hangs at startup if it does. E.g. you change shard count and try to restart a node while all other nodes are down - it will hang because it will wait for a read barrier without quorum available. I can change the contents of the message to something more high level (not sure what though) but I'd prefer for it to stay on INFO level.

Reply:
Not necessarily high level, but more descriptive would be good. You can just extend that one, e.g. "performing read barrier to check that topology is synchronized with local metadata"

Your doc above says: "
    // We avoid performing a read barrier if we're sure that our metadata stored in topology
    // is the same as local metadata."

But it seems to me the read barrier is unconditionally performed, even if local state is the same as the one stored in topology. So it seems either doc or code is incorrect. The former right?

Reply:
> You can just extend that one, e.g. "performing read barrier to check that topology is synchronized with local metadata"

How about:
```cpp
        slogger.info("raft topology: refreshing topology to check if it's synchronized with local metadata");
```